crates/sui-core/src/node_sync/node_follower.rs

```diff
@@ -2,797 +2,535 @@
+            // One item was processed from the db because the first one failed.
+            assert_eq!(result.items_from_db, 1);
 
-        let digests = TransactionEffectsDigest::random();
+            // At least 3 more were fetched from the stream - this number can't be checked exactly
+            // because the authority returns everything in the current batch even if fewer items
+            // were requested, and batch boundaries are non-deterministic.
+            assert!(result.items_from_stream >= 3 && result.items_from_stream <= 5);
+        }
 
-        assert!(!map.note_effects_digest(&byzantine, 1, threshold, &digests));
-        assert!(!map.note_effects_digest(&validator2, 1, threshold, &digests));
+        // test follower_process
+        {
+            debug!("testing follower_process");
+            let sync_store = new_sync_store();
+            let test_handler = TestNodeSyncHandler::new();
+
+            let metrics = GossipMetrics::new_for_tests();
+
+            let (_cancel_tx, cancel_rx) = tokio::sync::oneshot::channel();
+
+            let test_handler_clone = test_handler.clone();
+            let net_clone = net.clone();
+            let follower_task = tokio::task::spawn(async move {
+                let _ = timeout(
+                    Duration::from_secs(60),
+                    follower_process(
+                        test_handler_clone.break_after(2),
+                        sync_store,
+                        net_clone,
+                        5,
+                        &metrics,
+                        cancel_rx,
+                    ),
+                )
+                .await;
+            });
 
-        // double voting is rejected
-        assert!(!map.note_effects_digest(&byzantine, 1, threshold, &digests));
+            let intervals: Arc<Mutex<Vec<Duration>>> = Default::default();
+            let intervals_clone = intervals.clone();
+
+            let mut failure_rx = test_handler.failure_tx.subscribe();
+            let monitor_task = tokio::task::spawn(async move {
+                let start = Instant::now();
+                loop {
+                    failure_rx.recv().await.unwrap();
+                    test_handler.inc_break_after();
+                    let delay = Instant::now() - start;
+                    intervals_clone.lock().unwrap().push(delay);
+                    debug!("failure triggered {:?}", delay)
+                }
+            });
 
-        // final vote pushes us over.
-        assert!(map.note_effects_digest(&validator3, 1, threshold, &digests));
+            follower_task.await.unwrap();
+            monitor_task.abort();
 
-        // double vote doesn't result in false if we already exceeded threshold.
-        assert!(map.note_effects_digest(&byzantine, 1, threshold, &digests));
+            let ivs = intervals.lock().unwrap();
+            assert_eq!(ivs.len(), 7);
+            assert!(*ivs.last().unwrap() > Duration::from_secs(30));
```

Comment:
it's not very self-descriptive why `7` and `30`. let's add more comments if people can easily understand what may go wrong.

Reply:
commented